### PR TITLE
Diagnostics: Intro the public module name concept to hide support modules from clients

### PR DESCRIPTION
**Pull request by xymus**
Introduce the concept of a public facing module name to hide support modules from external clients. This name should be set on a support module and point to a public module for which it is an implementation detail. For example, BigKitCore’s public module name should be set to BigKit for diagnostics to refer to both modules as BigKit.

The public module name is used to refer to the module in diagnostics for most clients. Only clients that also imported the module identified by the public module name will see the public module name. Clients that don’t import the public module will still use the real name of the module. In practice, only other support modules under the public module should be in this position and see the real module names. External clients should all import only the public module.

The public module name is set with the flag `-public-module-name` and it's preserved in textual swiftinterfaces and binary swiftmodules.

Future work:
- The public module name should be used in documentation tools as well.
- Use clang’s `export_as` to get the same diagnostics behavior for imported clang modules.
- Use the knowledge of support to public module relationships for a coarse restriction on what client can directly import the support modules.
- Preserve the real module name in advanced diagnostics like `-Rmodule-api-import`.

Resolves rdar://134527933

### CODE REVIEW

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Comment by xymus** [comment]
@swift-ci Please smoke test

Updated the flag to be ignorable in swiftinterfaces.

**Review by tshortli — APPROVED** [review]

**Comment by xymus** [comment]
@swift-ci Please smoke test

File: include/swift/AST/Module.h [hunk 1]
```diff
@@ -504,6 +506,21 @@ class ModuleDecl
     ExportAsName = name;
   }
 
+  /// Public facing name for this module in diagnostics and documentation.
+  ///
+  /// This always returns a valid name as it defaults to the module name if
+  /// no public module name is set.
+  ///
+  /// If `mustBeVisible`, return the normal module name when the module
```
[the remark this answers is not in the document]
Comment (xymus) [reply include/swift/AST/Module.h]: I like it, I've changed it.

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Pull request merged**